test/test_env.py

```diff
@@ -900,6 +901,76 @@ def env_fn2(seed):
         env1.close()
         env2.close()
 
+    @pytest.mark.parametrize("batch_size", [(), (1,), (4,), (32, 5)])
+    @pytest.mark.parametrize("n_workers", [1, 2])
+    def test_parallel_env_reset_flag(self, batch_size, n_workers, max_steps=3):
+        env = ParallelEnv(
+            n_workers, lambda: CountingEnv(max_steps=max_steps, batch_size=batch_size)
+        )
+        env.set_seed(1)
+        action = env.action_spec.rand(env.batch_size)
+        action[:] = 1
+
+        for i in range(max_steps):
+            td = env.step(
+                TensorDict(
+                    {"action": action}, batch_size=env.batch_size, device=env.device
+                )
+            )
+            assert (td["done"] == 0).all()
+            assert (td["next"]["observation"] == i + 1).all()
+
+        td = env.step(
+            TensorDict({"action": action}, batch_size=env.batch_size, device=env.device)
+        )
+        assert (td["done"] == 1).all()
+        assert (td["next"]["observation"] == max_steps + 1).all()
+
+        _reset = torch.randint(
+            low=0, high=2, size=(*env.batch_size, 1), dtype=torch.bool
```

Comment:
why is the last dim 1? With the latest tensordict version, the dims of a tensor can match the batch size of a tensordict.

Reply:
I used the convention of the done flag which has shape `(*batch_size,1)`. I adopted this convetion for _reset also in all the PR files

Reply:
See here for example

Reply:
Or the one me and you did recently https://github.com/pytorch/rl/blob/main/torchrl/envs/vec_env.py#L1013

Reply:
For `"done"`, we want the last dim to be 1 because we do this kind of thing:
```python
value = reward + value * gamma * (1-done)
```
we want reward and done to have a last dim of 1 otherwise they will be casted to the size of `value` which is usually the output of a neural net like `nn.Sequential(..., nn.Linear(..., 1))`.
So either we squeeze value or we unsqueeze reward and done.
Squeezing requires more brain power from the users IMO, whereas unsqueezing can be hidden from them.

For `"_reset"` I don't think we need that

Reply:
There are then some choices we have to make in the `StepCounter`. Because this transoform in instantiation the "step_count" key which has shape (env.batch_dim,1)

So do we want to also make that key loose the 1? or we keep the 1 there and unsqueeze _reset?

Reply:
IMO StepCounter should have steps that match the tensordict batch size.
@riiswa

Reply:
ok so like it is now in the pr, i agree

Reply:
In the future it might be worth considering forcing one of the two convenctions for all tensordicts imo. I.e. all keys of shape (batch_size, 1) have to be squeezed. or the opposite. I understand that for reward and done one way might be more comfortable, but I find it confusing that some keys and specs are squeezed and others are not

Reply:
yep it's been under my radar for quite some time. It's a conflicting thing:
Before tensordict was designed such that you could not store tensors with a shape identical to the tensordict (you had to unsqueeze or it was done for your). 
That was surprising to many users, so we dropped that.
But as mentioned above, for reward and done, they interact so much with neural nets that it's pain to work with batched data that does not end with a singleton dimension. 
Maybe unsqueezing in the env is not the solution though, I don't know really. We could delegate that to the value functions and such. My main worry is that it may lead to silent errors, e.g.

```
done = torch.zeros(10)
next_value = torch.ones(10, 1)
value = done*next_value  # silently gives the wrong 10 x 10 tensor
```
it's not only a problem for us but also for the users. That's why I'm not sure we're doing them a favour by removing the unsqueeze. 

But I'm more than happy to keep the conversation going.

@tcbegley was involved in some of this refactoring, he may have a 2 cents to share